File: mobile/android/app/src/main/kotlin/com/codeiy/im/core/billing/RevenueCatBillingService.kt
package com.codeiy.im.core.billing

import android.app.Activity
import android.content.Context
import com.codeiy.im.BuildConfig
import com.codeiy.im.model.BillingInterval
import com.codeiy.im.model.PlanCode
import com.revenuecat.purchases.Package
import com.revenuecat.purchases.PurchaseParams
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.PurchasesConfiguration
import com.revenuecat.purchases.PurchasesTransactionException
import com.revenuecat.purchases.awaitLogIn
import com.revenuecat.purchases.awaitLogOut
import com.revenuecat.purchases.awaitOfferings
import com.revenuecat.purchases.awaitPurchase
import com.revenuecat.purchases.awaitRestore

class RevenueCatBillingService(context: Context) : BillingService {
    private val applicationContext = context.applicationContext
    private val apiKey = BuildConfig.REVENUECAT_ANDROID_PUBLIC_API_KEY.trim()
    private var currentUserId: String? = null
    private val packages = mutableMapOf<String, Package>()

    override val isConfigured: Boolean get() = apiKey.isNotEmpty()

    override suspend fun identify(userId: String) {
        require(userId.isNotBlank()) { "登录完成后才能购买套餐" }
        if (!isConfigured) throw BillingUnavailableException("支付尚未配置")
        if (currentUserId == userId) return
        if (Purchases.isConfigured) {
            Purchases.sharedInstance.awaitLogIn(userId)
        } else {
            Purchases.configure(
                PurchasesConfiguration.Builder(applicationContext, apiKey)
                    .appUserID(userId)
                    .build(),
            )
        }
        currentUserId = userId
        packages.clear()
    }

    override suspend fun clearIdentity() {
        packages.clear()
        currentUserId = null
        if (Purchases.isConfigured && !Purchases.sharedInstance.isAnonymous) {
            runCatching { Purchases.sharedInstance.awaitLogOut() }
        }
    }

    override suspend fun fetchPackages(): List<BillingPackageOption> {
        if (currentUserId == null) throw BillingUnavailableException("登录完成后才能购买套餐")
        val offering = Purchases.sharedInstance.awaitOfferings().all["default"]
            ?: throw BillingUnavailableException("当前没有可购买的套餐")
        return offering.availablePackages.mapNotNull { item ->
            val identity = packageIdentity(item.identifier) ?: return@mapNotNull null
            packages[item.identifier] = item
            BillingPackageOption(item.identifier, identity.first, identity.second, item.product.price.formatted)
        }
    }

    override suspend fun purchase(activity: Activity, packageId: String): BillingPurchaseResult {
        if (currentUserId == null) throw BillingUnavailableException("登录完成后才能购买套餐")
        val selected = packages[packageId] ?: throw BillingUnavailableException("所选套餐暂不可购买")
        return try {
            Purchases.sharedInstance.awaitPurchase(PurchaseParams.Builder(activity, selected).build())
            BillingPurchaseResult.Purchased
        } catch (error: PurchasesTransactionException) {
            if (error.userCancelled) BillingPurchaseResult.Cancelled else throw error
        }
    }

    override suspend fun restorePurchases() {
        if (currentUserId == null) throw BillingUnavailableException("登录完成后才能恢复购买")
        Purchases.sharedInstance.awaitRestore()
    }

    private fun packageIdentity(identifier: String): Pair<PlanCode, BillingInterval>? = when (identifier) {
        "plus_monthly" -> PlanCode.PLUS to BillingInterval.MONTHLY
        "plus_annual" -> PlanCode.PLUS to BillingInterval.ANNUAL
        "pro_monthly" -> PlanCode.PRO to BillingInterval.MONTHLY
        "pro_annual" -> PlanCode.PRO to BillingInterval.ANNUAL
        "max_monthly" -> PlanCode.MAX to BillingInterval.MONTHLY
        "max_annual" -> PlanCode.MAX to BillingInterval.ANNUAL
        else -> null
    }
}

File: mobile/ios/Core/Billing/RevenueCatBillingService.swift
import Foundation
import RevenueCat

@MainActor
final class RevenueCatBillingService: BillingService {
    private let configuration: BillingConfiguration
    private var currentUserID: UUID?
    private var packages: [String: RevenueCat.Package] = [:]

    var isConfigured: Bool { configuration.publicAPIKey != nil }

    init(configuration: BillingConfiguration = .current) {
        self.configuration = configuration
    }

    func identify(userID: UUID) async throws {
        guard let key = configuration.publicAPIKey else { throw BillingServiceError.notConfigured }
        guard currentUserID != userID else { return }
        let userIDString = userID.uuidString.lowercased()
        if Purchases.isConfigured {
            _ = try await Purchases.shared.logIn(userIDString)
        } else {
            Purchases.configure(withAPIKey: key, appUserID: userIDString)
        }
        currentUserID = userID
        packages.removeAll()
    }

    func clearIdentity() async {
        packages.removeAll()
        currentUserID = nil
        guard Purchases.isConfigured, !Purchases.shared.isAnonymous else { return }
        _ = try? await Purchases.shared.logOut()
    }

    func fetchPackages() async throws -> [BillingPackageOption] {
        guard currentUserID != nil else { throw BillingServiceError.notAuthenticated }
        let offerings = try await Purchases.shared.offerings()
        guard let offering = offerings.all[configuration.offeringID] else {
            throw BillingServiceError.offeringUnavailable
        }
        let supported = offering.availablePackages.compactMap { package -> BillingPackageOption? in
            guard let identity = Self.packageIdentity(package.identifier) else { return nil }
            packages[package.identifier] = package
            return BillingPackageOption(
                id: package.identifier,
                planCode: identity.plan,
                interval: identity.interval,
                localizedPrice: package.localizedPriceString
            )
        }
        return supported
    }

    func purchase(packageID: String) async throws -> BillingPurchaseResult {
        guard currentUserID != nil else { throw BillingServiceError.notAuthenticated }
        guard let package = packages[packageID] else { throw BillingServiceError.packageUnavailable }
        let result = try await Purchases.shared.purchase(package: package)
        return result.userCancelled ? .cancelled : .purchased
    }

    func restorePurchases() async throws {
        guard currentUserID != nil else { throw BillingServiceError.notAuthenticated }
        _ = try await Purchases.shared.restorePurchases()
    }

    private static func packageIdentity(_ identifier: String) -> (plan: PlanCode, interval: BillingInterval)? {
        let values: [String: (PlanCode, BillingInterval)] = [
            "plus_monthly": (.plus, .monthly), "plus_annual": (.plus, .annual),
            "pro_monthly": (.pro, .monthly), "pro_annual": (.pro, .annual),
            "max_monthly": (.max, .monthly), "max_annual": (.max, .annual),
        ]
        return values[identifier]
    }
}
